googlesource/googlesource.go

```diff
@@ -334,79 +292,64 @@ func (s *Source) Next(ctx context.Context) (sdk.Record, error) {
 }
 
 func fetchPos(s *Source, pos sdk.Position) {
-	s.positions = positions{positions: make(map[string]string)}
-	s.positions.lock = new(sync.Mutex)
-	s.positions.lock.Lock()
+	s.position = ""
 
-	err := json.Unmarshal(pos, &s.positions.positions)
+	err := json.Unmarshal(pos, &s.position)
 	if err != nil {
 		sdk.Logger(s.ctx).Info().Msg("Could not get position. Will start with offset 0")
 	}
-	s.positions.lock.Unlock()
 }
 
 func getTables(s *Source) (err error) {
 	if s.sourceConfig.Config.TableIDs == "" {
-		s.tables, err = s.listTables(s.sourceConfig.Config.ProjectID, s.sourceConfig.Config.DatasetID)
-		if err != nil {
-			sdk.Logger(s.ctx).Trace().Str("err", err.Error()).Msg("error found while listing table")
-		}
-	} else {
-		s.tables = strings.Split(s.sourceConfig.Config.TableIDs, ",")
+		sdk.Logger(s.ctx).Trace().Str("err", err.Error()).Msg("error found while listing table")
+		return fmt.Errorf("table ID blank")
 	}
+	s.table = s.sourceConfig.Config.TableIDs
 	return err
 }
 
 func (s *Source) runIterator() (err error) {
-	rowInput := make(chan readRowInput)
+	var wg sync.WaitGroup
 
 	if err = getTables(s); err != nil {
 		sdk.Logger(s.ctx).Trace().Str("err", err.Error()).Msg("error found while fetching tables. Need to stop proccessing ")
 		return err
 	}
 
-	var wg sync.WaitGroup
 	// Snapshot sync. Start were we left last
-	for _, tableID := range s.tables {
-		wg.Add(1)
-
-		s.tomb.Go(func() (err error) {
-			sdk.Logger(s.ctx).Trace().Msg(fmt.Sprintf("position %v : %v", tableID, s.getPosition()[tableID]))
-			return s.ReadGoogleRow(rowInput, s.records)
-		})
-		position := s.getPosition()
-		rowInput <- readRowInput{tableID: tableID, offset: position[tableID], positions: position, wg: &wg}
-	}
+	wg.Add(1)
+
+	rowInput := readRowInput{offset: s.position, positions: s.position, wg: &wg}
+	s.tomb.Go(func() (err error) {
+		sdk.Logger(s.ctx).Trace().Msg(fmt.Sprintf("position %v : %v", s.table, s.position))
+		return s.ReadGoogleRow(rowInput, s.records)
+	})
+
 	wg.Wait()
+
 	for {
 		select {
 		case <-s.tomb.Dying():
 			return s.tomb.Err()
 		case <-s.ticker.C:
 			sdk.Logger(s.ctx).Trace().Msg("ticker started ")
-
-			if err = getTables(s); err != nil {
-				sdk.Logger(s.ctx).Trace().Str("err", err.Error()).Msg("error found while fetching tables. Need to stop proccessing ")
-				return err
-			}
 			runCDCIterator(s, rowInput)
 		}
 	}
 }
```

Comment:
It feels like this code can be simplified more. For example, we do this in a number of places:
1. Create a wait group
2. Start something in a tomb (i.e. in a goroutine, concurrently)
3. Immediately after that we wait for the goroutine to finish.

In other words, we don't even need the goroutine.

Reply:
https://github.com/neha-Gupta1/conduit-connector-bigquery/compare/pullSingleTable...pullSingleTable-suggestion